In [187]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

## Data Partition

60 Train, 20 Validation, 20 Test
*or*
75 Train, 15 Validation, 15 Test

## KNN

k=1,3,5,7..odd

take odd numbers as there will be no 50-50 possibility

### Read the dataset into numpy array

In [188]:
data = np.genfromtxt("../Dataset/iris.csv", delimiter=',')
data[:10]

array([[5.1, 3.5, 1.4, 0.2, 0. ],
       [4.9, 3. , 1.4, 0.2, 0. ],
       [4.7, 3.2, 1.3, 0.2, 0. ],
       [4.6, 3.1, 1.5, 0.2, 0. ],
       [5. , 3.6, 1.4, 0.2, 0. ],
       [5.4, 3.9, 1.7, 0.4, 0. ],
       [4.6, 3.4, 1.4, 0.3, 0. ],
       [5. , 3.4, 1.5, 0.2, 0. ],
       [4.4, 2.9, 1.4, 0.2, 0. ],
       [4.9, 3.1, 1.5, 0.1, 0. ]])

In [189]:
my_list = data.tolist()

### Randomize the order of arrangement of data

In [190]:
import random
random.shuffle(my_list)

In [191]:
my_list[:10]

[[6.2, 3.4, 5.4, 2.3, 2.0],
 [5.4, 3.0, 4.5, 1.5, 1.0],
 [5.1, 3.3, 1.7, 0.5, 0.0],
 [7.7, 2.8, 6.7, 2.0, 2.0],
 [5.4, 3.4, 1.5, 0.4, 0.0],
 [5.7, 2.5, 5.0, 2.0, 2.0],
 [6.7, 3.1, 5.6, 2.4, 2.0],
 [7.7, 3.8, 6.7, 2.2, 2.0],
 [5.6, 2.5, 3.9, 1.1, 1.0],
 [6.0, 2.2, 5.0, 1.5, 2.0]]

### Add the lists to Train, Validation, and Test lists

In [192]:
train = []
valid = []
test = []

for item in my_list:
    test_val = random.uniform(0, 1.0)
    if test_val <= 0.7:
        train.append(item)
    elif test_val <= 0.85:
        valid.append(item)
    else:
        test.append(item)

In [193]:
len(train), len(valid), len(test)

(99, 29, 22)

# KNN Classification

In [227]:
def KNNAlgorithm(train, valid, K=5):
    correct = 0
    
    for validation in valid:
        distances_list = []
        samples = []
        counts = []
        for training in train:
            euclidean = sum([(p-q) ** 2 for p, q in zip(validation[:-1], training[:-1])])
            distances_list.append((euclidean, training))
        distances_list.sort(key = lambda dist : dist[0])
        samples = distances_list[:K]
    
        # for all samples upto K, find the max class from the data-set
        for sample in samples:
            data = sample[1]
            data = data[-1:]
            counts.append(data[0])
    
        classifier = max(set(counts), key = counts.count)
        
        if validation[-1:][0] == classifier:
            correct += 1
    
    return (correct/len(valid)) * 100

In [226]:
print(f"{1} = {KNNAlgorithm(train,valid,1)}")
print(f"{3} = {KNNAlgorithm(train,valid,3)}")

for i in range(5,100,5):
    print(f"{i} = {KNNAlgorithm(train,valid,i)}")

1 = 93.10344827586206
3 = 93.10344827586206
5 = 93.10344827586206
10 = 93.10344827586206
15 = 93.10344827586206
20 = 93.10344827586206
25 = 96.55172413793103
30 = 96.55172413793103
35 = 96.55172413793103
40 = 96.55172413793103
45 = 96.55172413793103
50 = 93.10344827586206
55 = 93.10344827586206
60 = 93.10344827586206
65 = 89.65517241379311
70 = 79.3103448275862
75 = 48.275862068965516
80 = 48.275862068965516
85 = 48.275862068965516
90 = 44.827586206896555
95 = 44.827586206896555


In [224]:
KNNAlgorithm(train, test, 30)

95.45454545454545

### K = 30 provides highest validation accuracy.
### This K value gives 95.5% in the test accuracy